# Задача 2

In [178]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn import datasets, linear_model
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split

In [179]:
X_train, X_test, y_train, y_test = train_test_split(diabetes.data, diabetes.target, test_size=0.8, random_state=0)

# Линейная регрессия

In [180]:
lireg=linear_model.LinearRegression().fit(X_train,y_train)

In [181]:
y_pred = lireg.predict(X_test)
mse = np.mean( (y_pred - y_test) ** 2)
score = lireg.score(X_test, y_test)

In [182]:
print("Коэффициенты: ", lireg.coef_)

Коэффициенты:  [  -6.92791214 -261.90320276  672.08752531  251.15776027  334.94666979
 -389.4469787  -468.47100766 -100.63001358  561.4986084   -29.06277576]


In [183]:
print("Линейная регрессия")
print("MSE %.2f" % mse)
print("Variance score: %.2f" % score)

Линейная регрессия
MSE 3185.85
Variance score: 0.44


# Ridge

In [184]:
model_ridge=linear_model.Ridge(alpha = 0.1).fit(X_train,y_train)

In [185]:
y_pred = model_ridge.predict(X_test)
mse = np.mean( (y_pred - y_test) ** 2)
score = model_ridge.score(X_test, y_test)

In [186]:
print("Коэффициенты: ", model_ridge.coef_)

Коэффициенты:  [  24.65392148 -148.73236167  429.98056486  197.27454614   40.48890331
  -62.21395316 -212.65256411  151.52688643  418.61874992   52.71695927]


In [187]:
print("Линейная регрессия")
print("MSE %.2f" % mse)
print("Variance score: %.2f" % score)

Линейная регрессия
MSE 3082.99
Variance score: 0.46


# Lasso

In [188]:
model_lasso=linear_model.Lasso(alpha = 0.1).fit(X_train,y_train)

In [189]:
y_pred = model_lasso.predict(X_test)
mse = np.mean( (y_pred - y_test) ** 2)
score = model_lasso.score(X_test, y_test)

In [190]:
print("Коэффициенты: ", model_lasso.coef_)

Коэффициенты:  [  -0.         -195.18747079  622.29787523  193.53965662   -0.
  -56.04919911 -223.48292811    0.          650.69976584   -0.        ]


In [191]:
print("Линейная регрессия")
print("MSE %.2f" % mse)
print("Variance score: %.2f" % score)

Линейная регрессия
MSE 3102.64
Variance score: 0.46


# ElasticNet

In [192]:
model_elastic==linear_model.ElasticNet(alpha = 0.1, l1_ratio = 0.9).fit(X_train,y_train)

False

In [193]:
y_pred = model_elastic.predict(X_test)
mse = np.mean( (y_pred - y_test) ** 2)
score = model_elastic.score(X_test, y_test)

In [194]:
print("Коэффициенты: ", model_elastic.coef_)

Коэффициенты:  [ 38.22620231  -4.75842324 141.71226944  91.65587511  49.18799502
  26.77620653 -99.38964289 112.13425869 156.82861294  69.22724482]


In [195]:
print("Линейная регрессия")
print("MSE %.2f" % mse)
print("Variance score: %.2f" % score)

Линейная регрессия
MSE 4004.45
Variance score: 0.30


# Вывод

Mean Square Error — средний квадрат ошибки определения какой-либо величины, квадратный корень из MSE есть среднеквадратическое отклонение определяемой величины от её математического ожидания. Сравнивая значения MSE, можно сказать, что Ridge Regression(со значением MSE 3082.99) показала себя наилучшим образом, а ElasticNet же (со значением 4004.45), наоборот, худшим.